Support Glue catalog in Iceberg connector

alexjo2144 opened the pull request:
Authored by @jackye1995, based off of https://github.com/trinodb/trino/pull/10151

I still need to address the testing and figure out what is possible to run on the Trino automation.

findepi commented:
@alexjo2144 can you please squash commits before re-review? it's been a while, so i need to start reading anew anyway.

alexjo2144 commented:
Squashed. I left a copy of the history on a separate branch in case it's useful to see what I changed from Jack's initial PR https://github.com/alexjo2144/trino/tree/alexjo/glue-iceberg-catalog-full-history

alexjo2144 commented:
@findepi I took a stab at adapted Jack's TestIcebergGlueConnectorSmokeTest to work on CI. PTAL

alexjo2144 commented:
Pushed another commit with some more Catalog tests, and one which should fix https://github.com/trinodb/trino/issues/10902

alexjo2144 commented:
I am still working on the inverse problem, where reading the Iceberg Glue information schema fails if Hive tables are present. I'll try to have that fix and a test up later today.

alexjo2144 commented:
> I am still working on the inverse problem, where reading the Iceberg Glue information schema fails if Hive tables are present. I'll try to have that fix and a test up later today.

Just pushed an update for that in the last commit along with some tests for that case. The fixups were getting a little out of hand so I did squash them.

alexjo2144 commented:
One other thing I noticed while I was writing these tests though. `SHOW TABLES FROM iceberg.information_schema` fails on Glue but passes on the Hive catalog.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueToTrinoConverter.java:
```diff
@@ -66,10 +66,17 @@ public static Database convertDatabase(com.amazonaws.services.glue.model.Databas
-        // TODO (https://github.com/trinodb/trino/issues/10902) glueTable.getStorageDescriptor() is an optional field in Glue
-        requireNonNull(glueTable.getStorageDescriptor(), () -> format("Table StorageDescriptor is null for table %s.%s (%s)", dbName, glueTable.getName(), glueTable));
+        if (glueTable.getStorageDescriptor() == null) {
+            return Optional.empty();
```
Will this cause "table not found" to be returned when someone tries to access such a table?
This is confusing: user sees a table in SHOW TABLES, but then we tell them it doesn't exist.

alexjo2144 replied:
Yeah, it will. I have a test showing that here https://github.com/trinodb/trino/pull/10845/files#diff-07320a12f78f0c77af590dd6a46907f3c882bdef8e140b01a27ae8dd2b3bd718R155

I guess the other options would be to throw something like a `NotAHiveTableException` and handle that as a special case in the callers

alexjo2144 commented:
@findepi I pushed an update with a fix for table redirections, and some tests with a Glue metastore that has Hive and Iceberg tables in the same namespace.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+                return namespaces.stream()
+                        .flatMap(glueNamespace -> {
+                            try {
+                                GetTablesRequest getTablesRequest = new GetTablesRequest().withDatabaseName(glueNamespace);
```
can you set `GetTablesRequest` without `withDatabaseName` to get all tables from all namespaces?

alexjo2144 replied:
No, databaseName is required unfortunately

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+                Optional.empty(),
+                Optional.empty());
+        operations.initializeFromMetadata(metadata);
+        return new BaseTable(operations, quotedTableName(table));
```
would `io.trino.plugin.iceberg.IcebergUtil#getIcebergTableWithMetadata` be applicable here?

if not, we should move that method to `TrinoHiveCatalog`

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+    }
+
+    @Override
+    public Transaction newCreateTableTransaction(
```
Looks identical to `io.trino.plugin.iceberg.catalog.hms.TrinoHiveCatalog#newCreateTableTransaction`.
should this be unified?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+            newTableCreated = true;
+
+            deleteTable(from.getSchemaName(), from.getTableName());
+            oldTableDeleted = true;
```
redundant store

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+                    .withName(from.getTableName());
+            catalogId.ifPresent(getTableRequest::setCatalogId);
+            com.amazonaws.services.glue.model.Table table = stats.getGetTable().call(() -> glueClient.getTable(getTableRequest).getTable());
+            TableInput tableInput = new TableInput()
+                    .withName(to.getTableName())
```
`io.trino.plugin.iceberg.catalog.hms.TrinoHiveCatalog#newCreateTableTransaction` also creates table copy, ie a new TableInput from `com.amazonaws.services.glue.model.Table`
yet, the logic differs heavily, for no apparent reason

have a method that unifies the logic, as it should stay in sync.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+                .withDatabaseName(schema)
+                .withName(table);
+        catalogId.ifPresent(deleteNewTableRequest::setCatalogId);
+        stats.getDropTable().call(() -> glueClient.deleteTable(deleteNewTableRequest));
```
`stats.getDeleteTable` (stats method out of sync with actual call)

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+    }
+
+    @Override
+    public void updateTableComment(ConnectorSession session, SchemaTableName schemaTableName, Optional<String> comment)
```
same as in `io.trino.plugin.iceberg.catalog.hms.TrinoHiveCatalog#updateTableComment`
it's not metastore-specific.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+    }
+
+    @Override
+    public String defaultTableLocation(ConnectorSession session, SchemaTableName schemaTableName)
```
Extract common part of this & `io.trino.plugin.iceberg.catalog.hms.TrinoHiveCatalog#defaultTableLocation` into a method, as the logic should stay in sync.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+    @Override
+    public void createView(ConnectorSession session, SchemaTableName schemaViewName, ConnectorViewDefinition definition, boolean replace)
+    {
+        throw new TrinoException(NOT_SUPPORTED, "createView is not supported for Iceberg Glue catalogs");
```
Iceberg connector support Trino views.
Please implement this (might be as follow-up)

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,520 @@
+    public void createMaterializedView(ConnectorSession session, SchemaTableName schemaViewName, ConnectorMaterializedViewDefinition definition,
+            boolean replace, boolean ignoreExisting)
+    {
+        throw new TrinoException(NOT_SUPPORTED, "createMaterializedView is not supported for Iceberg Glue catalogs");
```
Iceberg connector support materialized views.
Please implement this (might be as follow-up)

alexjo2144 replied:
https://github.com/trinodb/trino/issues/11268

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java:
```diff
@@ -0,0 +1,226 @@
+            return stats.getGetTable().call(() -> glueClient.getTable(getTableRequest).getTable());
+        }
+        catch (EntityNotFoundException e) {
+            throw new TableNotFoundException(getSchemaTableName());
```
Worth attaching `e`?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/AbstractMetastoreTableOperations.java:
```diff
@@ -113,7 +112,7 @@ protected void commitNewTable(TableMetadata metadata)
                 e.addSuppressed(ex);
             }
-            throw new TrinoException(ICEBERG_COMMIT_ERROR, "Failed to commit new table to the Iceberg Catalog", e);
+            throw e;
```
is it a revert of the change in first commit?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java:
```diff
@@ -95,7 +92,7 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
                     e.addSuppressed(ex);
                 }
-                throw new TrinoException(ICEBERG_COMMIT_ERROR, format("Failed to commit to the Iceberg Catalog for table %s.%s", database, tableName), e);
+                throw e;
```
is it a revert of the change in first commit?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -128,9 +129,9 @@
-import static io.trino.plugin.iceberg.TrinoHiveCatalog.DEPENDS_ON_TABLES;
 import static io.trino.plugin.iceberg.TypeConverter.toIcebergType;
 import static io.trino.plugin.iceberg.TypeConverter.toTrinoType;
+import static io.trino.plugin.iceberg.catalog.hms.TrinoHiveCatalog.DEPENDS_ON_TABLES;
```
TrinoHiveCatalog is not HMS-specific, so should not be in hms package.

alexjo2144 replied:
Do you have a better name for the package that has the HMS and File Hive Metastore classes?

findepi replied:
Maybe just put it in `.catalog`?
Does  the separation into `.catalog.hms` buy us much?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java:
```diff
@@ -303,11 +306,7 @@ public boolean dropTable(ConnectorSession session, SchemaTableName schemaTableNa
-                table.properties().containsKey(WRITE_METADATA_LOCATION)) {
-            throw new TrinoException(NOT_SUPPORTED, "Table " + schemaTableName + " contains Iceberg path override properties and cannot be dropped from Trino");
-        }
+        validateTableCanBeDropped(table, schemaTableName);
```
Extract `validateTableCanBeDropped` extraction to a prep commit.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalogFactory.java:
```diff
@@ -0,0 +1,88 @@
+        this.trinoVersion = requireNonNull(nodeVersion, "trinoVersion is null").toString();
+        requireNonNull(config, "config is null");
+        this.isUniqueTableLocation = config.isUniqueTableLocation();
+        this.isUsingSystemSecurity = securityConfig.getSecuritySystem() == SYSTEM;
```
any reason `config` is `rnn`d but `securityConfig` is not?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalogFactory.java:
```diff
@@ -0,0 +1,88 @@
+        requireNonNull(config, "config is null");
+        this.isUniqueTableLocation = config.isUniqueTableLocation();
+        this.isUsingSystemSecurity = securityConfig.getSecuritySystem() == SYSTEM;
+        this.deleteSchemaLocationsFallback = requireNonNull(hiveConfig).isDeleteSchemaLocationsFallback();
```
add msg to `rnn`, otherwise it's useless, as the NPE would fly anyway.

(and counter-productive in Java 17 (JEP 358))

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java:
```diff
@@ -0,0 +1,54 @@
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public abstract class BaseIcebergConnectorSmokeTest
```
Extract BaseIcebergConnectorSmokeTest extraction to a separate commit.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseTestTrinoCatalog.java:
```diff
@@ -0,0 +1,191 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.assertNotEquals;
+
+public abstract class BaseTestTrinoCatalog
```
`BaseTrinoCatalogTest`

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestTrinoHiveCatalog.java:
```diff
@@ -0,0 +1,98 @@
+    public void testCreateNamespaceWithLocation()
+    {
+        assertThatThrownBy(super::testCreateNamespaceWithLocation)
+                .hasMessageContaining("Database cannot be created with a location set");
```
yikes, why?
it sounds like an important feature

alexjo2144 replied:
It is supported on HMS and Glue but not FileHiveMetastore

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestTrinoHiveCatalog.java:
```diff
@@ -0,0 +1,98 @@
+        HdfsConfiguration configuration = new HiveHdfsConfiguration(new HdfsConfigurationInitializer(config), ImmutableSet.of());
+        hdfsEnvironment = new HdfsEnvironment(configuration, config, new NoHdfsAuthentication());
+
+        tempDir = Files.createTempDirectory("test_iceberg_v2");
```
test_iceberg_v2 ?

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestTrinoGlueCatalog.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Test
+    public void testDefaultLocation()
```
Can this be in `BaseTestTrinoCatalog`?

alexjo2144 replied:
This tests the `hive.metastore.glue.default-warehouse-dir` config, which is Glue specific

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestTrinoHiveCatalog.java:
```diff
@@ -0,0 +1,98 @@
+import static io.trino.plugin.hive.metastore.file.FileHiveMetastore.createTestingFileHiveMetastore;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestTrinoHiveCatalog
```
This class tests a pre-existing functionality. I'd extract it to a separate commit.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/TrinoCatalog.java:
```diff
@@ -47,7 +47,7 @@
     List<String> listNamespaces(ConnectorSession session);
 
-    boolean dropNamespace(ConnectorSession session, String namespace);
+    void dropNamespace(ConnectorSession session, String namespace);
```
This should be a prep commit. Can you please reorder commits?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -780,14 +780,13 @@ private ConnectorTableMetadata getTableMetadata(ConnectorSession session, Schema
-                            .setComment(Optional.ofNullable(column.doc()))
-                            .build();
-                })
+                .map(column ->
```
extract cleanups to a separate PR for quicker merge

alexjo2144 replied:
https://github.com/trinodb/trino/pull/11233

nineinchnick commented on .github/workflows/ci.yml:
```diff
@@ -411,6 +411,56 @@ jobs:
             **/surefire-reports/TEST-*.xml
           retention-days: ${{ env.TEST_REPORT_RETENTION_DAYS }}
 
+  test-iceberg:
```
I'm collapsing memsql and bigdata into the test job in #10323, see https://github.com/trinodb/trino/pull/10323/commits/9fe62826f27c29953ccc0f539ad5f9fb0c92bfa2

findepi replied:
thanks @nineinchnick . how should we adapt here? 
or, should we continue with what we have and revisit once that PR merged?

nineinchnick replied:
I'd advise moving this as a new step in the `tests` job like in the commit I pasted above.

nineinchnick replied:
#10323 has just been merged, lmk if you'd like me to provide an example how to pull this into the `test` job.

nineinchnick commented on .github/workflows/ci.yml:
```diff
@@ -411,6 +411,56 @@ jobs:
+          S3_BUCKET: "presto-ci-test"
+        run: |
+          if [ "${AWS_ACCESS_KEY_ID}" != "" ]; then
+            $MAVEN test ${MAVEN_TEST} -pl :trino-iceberg -P test-all -Ds3.bucket=${S3_BUCKET}
```
The difference here is activating an additional profile `test-all`. Do we still want to run these tests twice?

findepi replied:
the glue related tests are run once, the other tests are run twice.
no, we don't want  that :)

nineinchnick commented on .github/workflows/ci.yml:
```diff
@@ -411,6 +411,56 @@ jobs:
+          AWS_REGION: "us-east-2"
+          S3_BUCKET: "presto-ci-test"
+        run: |
+          if [ "${AWS_ACCESS_KEY_ID}" != "" ]; then
```
This should be moved as an additional step to the `test` job and it should be conditional, by adding `if:  env.AWS_ACCESS_KEY_ID != ''`. This will allow running the whole ci in forks without this step failing.

nineinchnick replied:
See how I did this for memsql and bigquery in https://github.com/trinodb/trino/pull/10323/commits/9fe62826f27c29953ccc0f539ad5f9fb0c92bfa2

findepi replied:
>  This will allow running the whole ci in forks without this step failing.

that's the obvious upside. the downside is that a ci typo disables a test.
anyway; pre-existing, not of a concern for this PR specifically

alexjo2144 commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueToTrinoConverter.java:
```diff
@@ -68,11 +71,26 @@ public static Database convertDatabase(com.amazonaws.services.glue.model.Databas
         StorageDescriptor sd = glueTable.getStorageDescriptor();
+        if (sd == null) {
+            if (isIcebergTable(tableParameters) || isDeltaLakeTable(tableParameters)) {
+                // Iceberg and Delta Lake tables do not use the StorageDescriptor field, but we need to return a Table so the caller can check that that
```
Delta doesn't use the storage format field, but it is usually set to some dummy value

findepi commented:
@alexjo2144 can you please extract [Handle missing StorageDescriptor in Hive Glue tables](https://github.com/trinodb/trino/pull/10845/commits/5607505f0e30fcbc8908b23275df602f30563022) to a separate PR?
once it's merged, i will update https://github.com/trinodb/trino/pull/10881 

doing otherwise would cause breakage for the master CI runs, right?

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java:
```diff
@@ -0,0 +1,226 @@
+                    .withDatabaseName(database)
+                    .withTableInput(tableInput);
+            catalogId.ifPresent(createTableRequest::setCatalogId);
+            stats.getCreateTable().call(() -> glueClient.createTable(createTableRequest));
```
It does not, doesn't look like we're retrying `glueClient` calls anywhere at the moment.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java:
```diff
@@ -0,0 +1,226 @@
+        }
+        catch (Exception e) {
+            try {
+                cleanupMetadataLocation(newMetadataLocation);
```
Aren't these kinds of asymmetrical network issues potential problems everywhere though? Seems like worrying about that would open up a very large can or worms

electrum replied:
We should not cleanup on generic commit failure. Storage cleanup happens later through GC.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java:
```diff
@@ -0,0 +1,226 @@
+                    .withLastAnalyzedTime(table.getLastAnalyzedTime())
+                    .withPartitionKeys(table.getPartitionKeys())
+                    .withRetention(table.getRetention())
+                    .withStorageDescriptor(table.getStorageDescriptor())
```
Yeah, I think I'd rather take out all of these `withX` calls. None of them should be applicable.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java:
```diff
@@ -0,0 +1,226 @@
+        shouldRefresh = true;
+    }
+
+    private boolean isPrestoView(Table table)
+    {
+        return "true".equals(table.getParameters().get(PRESTO_VIEW_FLAG));
+    }
+
+    private boolean isHiveOrPrestoView(Table table)
+    {
+        return table.getTableType().equals(TableType.VIRTUAL_VIEW.name());
+    }
+
+    private boolean isIcebergTable(Table table)
+    {
+        return ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(table.getParameters().get(TABLE_TYPE_PROP));
+    }
```
They're different `Table` classes, but I can try giving them common code.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperationsProvider.java:
```diff
@@ -0,0 +1,61 @@
+    {
+        TrinoGlueCatalog glueCatalog = (TrinoGlueCatalog) catalog;
+        return new GlueIcebergTableOperations(
+                glueCatalog.getGlueClient(),
+                glueCatalog.getStats(),
+                glueCatalog.getCatalogId(),
```
Yeah, I agree. A wrapper is a bit of a pain just since the GlueClient API is so large.

What do you think about taking `catalogId` support out for now, fail if it is explicitly set, until we can add testing and a less error-prone approach?

findepi replied:
> What do you think about taking catalogId support out for now

i'm fine.

> fail if it is explicitly set

just remove config option (and have followup issue)

> until we can add testing and a less error-prone approach?

describe the "TAP" in the issue

alexjo2144 replied:
> just remove config option

The config options are being reused from the Hive connector's `GlueHiveMetastoreConfig` right now, so to do that I'd have to make a copy of the config class.

findepi replied:
oh, ok, fine. Let's have a check for now then

RameshByndoor commented:
@alexjo2144 Thanks, we are waiting to see this feature released. Can you please help with when is this going to be available?

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java:
```diff
@@ -0,0 +1,226 @@
+        Table table = getTable();
+
+        if (isPrestoView(table) && isHiveOrPrestoView(table)) {
+            // this is a Presto Hive view, hence not a table
+            throw new TableNotFoundException(getSchemaTableName());
```
This is how the existing implementation with Hive works now. i think catching Views before this point would require a fair bit of refactoring. 

https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractMetastoreTableOperations.java#L146

alexjo2144 replied:
https://github.com/trinodb/trino/issues/11267

alexjo2144 commented:
Just pushed a rebase with the comments addressed and merge conflicts resolved. PTAL

findepi commented:
`io.trino.plugin.iceberg.TestIcebergPlugin#testRecordingMetastore` failure looks related.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestSharedGlueMetastore.java:
```diff
@@ -0,0 +1,233 @@
+import static org.testng.Assert.assertEquals;
+
+/**
+ * Tests metadata operations on a schema which has a mix of Hive and Iceberg tables.
```
We already have TestIcebergHiveTablesCompatibility, TestIcebergHiveViewsCompatibility and TestIcebergRedirectionToHive.

I like the fact it's runnable from IDE (if you have the creds). But i still would prefer making it into a product test environment so that we can directly run the above mentioned tests against HMS and Glue alike.
Of course this is best done when Iceberg Glue supports views and MVs, so ok to be as a followup.

findepi commented:
I updated https://github.com/trinodb/trino/pull/10881 with current state of this PR, so let me cancel the build here.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/IcebergQueryRunner.java:
```diff
@@ -81,9 +101,9 @@ public static DistributedQueryRunner createIcebergQueryRunner(
     }
 
-    public static Builder builder()
+    public static Builder builder(Optional<String> schema)
```
cc @aczajkowski

alexjo2144 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergPlugin.java:
```diff
@@ -99,6 +101,16 @@ public void testRecordingMetastore()
                 .shutdown();
+
+        // recording with glue
+        assertThatThrownBy(() -> factory.create(
+                "test",
+                Map.of(
+                        "iceberg.catalog.type", "glue",
+                        "hive.metastore.glue.region", "us-east-2",
+                        "hive.metastore-recording-path", "/tmp"),
+                new TestingConnectorContext()))
+                .hasMessageContaining("Configuration property 'hive.metastore-recording-path' was not used");
```
FWIW, the Hive plugin allows it. https://github.com/trinodb/trino/blob/master/plugin/trino-hive-hadoop2/src/test/java/io/trino/plugin/hive/TestHivePlugin.java#L143

alexjo2144 replied:
Never mind, I see what you're saying. You're right

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergUtil.java:
```diff
@@ -0,0 +1,20 @@
+
+import static org.apache.hadoop.hive.metastore.TableType.EXTERNAL_TABLE;
+
+public class GlueIcebergUtil
```
No, but I'm still running a 2020 build. I can resolve that warning either way though. I thought there used to be a style check for that.

findepi replied:
>  thought there used to be a style check for that.

i don't think so,

> No, but I'm still running a 2020 build. 

version probably doesn't matter, this does: https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#additional-ide-configuration

alexjo2144 commented:
AC and created related Issues, thanks!

findepi commented:
Thanks @alexjo2144 please rebase, there is a conflict.

alexjo2144 commented on .github/workflows/ci.yml:
```diff
@@ -458,6 +458,14 @@ jobs:
         if: matrix.modules == 'plugin/trino-bigquery' && env.BIGQUERY_CASE_INSENSITIVE_CREDENTIALS_KEY != ''
         run: |
           $MAVEN test ${MAVEN_TEST} -pl :trino-bigquery -Pcloud-tests-case-insensitive-mapping -Dbigquery.credentials-key="${BIGQUERY_CASE_INSENSITIVE_CREDENTIALS_KEY}"
+      - name: Iceberg Glue Catalog Tests
```
@nineinchnick can you take a look at this? This step didn't run on the CI where I was expecting it to

https://github.com/trinodb/trino/runs/5423189712?check_suite_focus=true

alexjo2144 replied:
Ah, I think I see it. `env.AWS_ACCESSKEY` should be `env.AWS_ACCESS_KEY_ID`

alexjo2144 commented:
The tests found a bug in CachingHiveMetastore's cache invalidation, putting together a fix for that now.

findepi commented on lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java:
```diff
@@ -346,7 +348,9 @@ public void clear()
-                throw new UnsupportedOperationException();
+                return tokens.entrySet().stream()
+                        .map(tokenEntry -> new AbstractMap.SimpleEntry<K, V>(tokenEntry.getKey(), dataCacheMap.get(tokenEntry.getValue())))
+                        .collect(toImmutableSet());
```
`Map.entrySet` is a view, not a copy

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:
```diff
@@ -604,6 +605,14 @@ private void invalidateTableStatisticsCache(String databaseName, String tableNam
 
+    private void invalidateTablesWithParameterCache(String databaseName, String tableName)
+    {
+        tablesWithParameterCache.asMap().entrySet().stream()
+                .filter(cacheEntry -> cacheEntry.getKey().getDatabaseName().equals(databaseName) && cacheEntry.getValue().contains(tableName))
```
iterate over keySet

alexjo2144 replied:
Sure. Updated to iterate over keySet and then do a lookup for each key.

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:
```diff
@@ -604,6 +605,17 @@ private void invalidateTableStatisticsCache(String databaseName, String tableNam
+        tablesWithParameterCache.asMap().keySet().stream()
+                .filter(cacheKey -> cacheKey.getDatabaseName().equals(databaseName))
+                .filter(cacheKey -> {
+                    List<String> cacheValue = tablesWithParameterCache.getIfPresent(cacheKey);
```
Shouldn't we implement `entrySet()` in `EvictableCache` and use that to avoid looking up each item?

findepi replied:
@electrum see https://github.com/trinodb/trino/pull/10845#discussion_r821058122

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java:
```diff
@@ -95,7 +98,7 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
                     e.addSuppressed(ex);
                 }
-                throw e;
+                throw new TrinoException(ICEBERG_COMMIT_ERROR, format("Failed to commit to table %s.%s", database, tableName), e);
```
Can simplify to
```java
throw new TrinoException(ICEBERG_COMMIT_ERROR, "Failed to commit to table: " + table.getSchemaTableName(), e);
```
Since `SchemaTableName` already formats nicely for error messages

alexjo2144 replied:
Can't log with `table` because it's defined in the try block, should be able to clean it up separately along with the other comment:

> This catch block is actually in the wrong place. The call most likely to fail is the replaceTable() call below, so this should really be on the outside. Though we could fix this separately.

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java:
```diff
@@ -95,7 +98,7 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
                     e.addSuppressed(ex);
                 }
-                throw e;
+                throw new TrinoException(ICEBERG_COMMIT_ERROR, format("Failed to commit to table %s.%s", database, tableName), e);
```
This catch block is actually in the wrong place. The call most likely to fail is the `replaceTable()` call below, so this should really be on the outside. Though we could fix this separately.

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:
```diff
@@ -403,4 +407,15 @@ public static Transaction newCreateTableTransaction(TrinoCatalog catalog, Connec
+                table.properties().containsKey(WRITE_NEW_DATA_LOCATION) ||
+                table.properties().containsKey(WRITE_METADATA_LOCATION) ||
+                table.properties().containsKey(WRITE_DATA_LOCATION)) {
+            throw new TrinoException(NOT_SUPPORTED, "Table " + schemaTableName + " contains Iceberg path override properties and cannot be dropped from Trino");
```
It's better to put the variable part of the message at the end, so it's easier to search for error strings:
```java
throw new TrinoException(NOT_SUPPORTED, "Table contains Iceberg path override properties and cannot be dropped from Trino: " + table);

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:
```diff
@@ -403,4 +407,15 @@ public static Transaction newCreateTableTransaction(TrinoCatalog catalog, Connec
         return catalog.newCreateTableTransaction(session, schemaTableName, schema, partitionSpec, targetPath, propertiesBuilder.buildOrThrow());
     }
+
+    public static void validateTableCanBeDropped(Table table, SchemaTableName schemaTableName)
```
No need to pass `SchemaTableName` here, since we construct `Table` objects with the correct name.

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,452 @@
+    @Override
+    public void createNamespace(ConnectorSession session, String namespace, Map<String, Object> properties, TrinoPrincipal owner)
+    {
+        verify(
```
This should be `checkArgument`. We can split these
```java
checkArgument(owner.getType() == PrincipalType.USER, "Owner type must be USER");
checkArgument(owner.getName().equals(session.getUser()), "Explicit schema owner is not supported");

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,452 @@
+                    .withDatabaseName(from.getSchemaName())
+                    .withName(from.getTableName());
+            com.amazonaws.services.glue.model.Table table = stats.getGetTable().call(() -> glueClient.getTable(getTableRequest).getTable());
+            TableInput tableInput = getTableInput(
```
No need to wrap

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,452 @@
+            newTableCreated = true;
+            deleteTable(from.getSchemaName(), from.getTableName());
+        }
+        catch (Exception e) {
```
Avoid catching `Exception` since it can hide `InterruptedException`. Instead, catch `RuntimeException` and any necessary checked exceptions.

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,452 @@
+                        HIVE_DATABASE_LOCATION_ERROR,
+                        format("Schema '%s' location cannot be determined. " +
+                                        "Either set the 'location' property when creating the schema, or set the 'hive.metastore.glue.default-warehouse-dir' " +
+                                        "config option.",
```
config option -> catalog property

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -0,0 +1,452 @@
+
+    @Override
+    public void createMaterializedView(ConnectorSession session, SchemaTableName schemaViewName, ConnectorMaterializedViewDefinition definition,
+            boolean replace, boolean ignoreExisting)
```
Don't wrap

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalogFactory.java:
```diff
@@ -0,0 +1,75 @@
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.tableOperationsProvider = requireNonNull(tableOperationsProvider, "tableOperationsProvider is null");
+        requireNonNull(glueConfig, "glueConfig is null");
+        verify(glueConfig.getCatalogId().isEmpty(), "catalogId configuration is not supported");
```
`checkArgument`

Why don't we support this?

findepi replied:
It was initially done in a very error-prone way (scattered all over the calls), and without test coverage. There is a separate issue to re-add this.

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/IcebergHiveMetastoreCatalogModule.java:
```diff
@@ -27,5 +30,8 @@ protected void setup(Binder binder)
         install(new ThriftMetastoreModule());
         binder.bind(IcebergTableOperationsProvider.class).to(HiveMetastoreTableOperationsProvider.class).in(Scopes.SINGLETON);
+        binder.bind(TrinoCatalogFactory.class).to(TrinoHiveCatalogFactory.class).in(Scopes.SINGLETON);
+        binder.bind(IcebergCatalogModule.MetastoreValidator.class).asEagerSingleton();
```
Import `MetastoreValidator` directly

electrum commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestTrinoHiveCatalogTest.java:
```diff
@@ -70,7 +71,7 @@ protected TrinoCatalog createTrinoCatalog(boolean useUniqueTableLocations)
         return new TrinoHiveCatalog(
                 new CatalogName("catalog"),
-                metastore,
+                memoizeMetastore(metastore, 1000),
```
Why?

findepi commented on .github/workflows/ci.yml:
```diff
@@ -459,6 +459,14 @@ jobs:
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESSKEY }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRETKEY }}
+          S3_BUCKET: "presto-ci-test"
```
it's better to have env-specific setup clearly in one place, so (bucket name, access keys) should be kept together

findepi commented on .github/workflows/ci.yml:
```diff
@@ -459,6 +459,14 @@ jobs:
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRETKEY }}
+          S3_BUCKET: "presto-ci-test"
+        if: matrix.modules == 'plugin/trino-iceberg' && (env.AWS_ACCESS_KEY_ID != '' || env.AWS_SECRET_ACCESS_KEY != '')
+        run: |
```
`|` allows you to avoid quoting (i.e. value could be interpreted as something else than a string)

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:
```diff
@@ -237,7 +236,7 @@ public static IcebergFileFormat getFileFormat(Table table)
     }
 
-    private static String quotedTableName(SchemaTableName name)
+    public static String quotedTableName(SchemaTableName name)
```
https://github.com/trinodb/trino/pull/10845/files#diff-0c09267905a13855768761575aee308f9a1773e6f3d59c6347de78fa8dbc1c02R254

alexjo2144 commented:
Comments addressed, thanks for the review David

alexjo2144 commented:
Last force push was just check style. I missed some `buildOrThrow()`s, hopefully it's happy now

findepi commented:
I updated https://github.com/trinodb/trino/pull/10881 to have CI run with secrets too.

electrum commented:
The tests failed due to needing the `AWS_REGION` environment variable set. I pushed a commit to fix that

alexjo2144 commented:
Taking a look at the test failure, it passes locally...

alexjo2144 commented:
I was able to figure out why that test is failing, I'm not sure if it's worth fixing here or just taking out the test.

The problem happens when there's an information_schema query like `SELECT * FROM hive.information_schema.tables WHERE table_name = 'region'` and there's an Iceberg table named `region` somewhere in one of the Glue schemas. `InformationSchemaPageSource` eventually calls `HiveMetadata#getTableHandle` for the Iceberg region table, and that fails since it's not a Hive table and redirections are not enabled. 

I had a hard time reproducing this since `InformationSchemaMetadata` does something different if there are > 100 tables which doesn't result in a `getTableHandle` call, and so it doesn't fail.

@electrum @findepi how do you two want to handle this?

electrum commented:
I think this is a bug in the engine. The code in `isExistingRelation()` should ignore any `TrinoException` when calling `getRedirectionAwareTableHandle()`. It probably needs to return an `Optional<Boolean>` and fallback to the normal listing code when not present, so that we return consistent results with and without a filter.

Related, using `getRedirectionAwareTableHandle()` for `listTables()` is actually not correct, since connectors might synthesize tables that do not appear in the `listTables()` call, thus leading to inconsistent results depending on the filter condition. I think we solve this by rethinking all of the listing and metadata APIs.

For this PR, let's remove this test and file an issue.

electrum commented:
I decided to fix it and pushed a new commit to #10881

electrum commented:
See https://github.com/trinodb/trino/pull/10881/commits/5875beb6f148c5cf48ffcc855bb2747574ddeffc

electrum commented:
See https://github.com/trinodb/trino/pull/10881/commits/1ffd25550bc07c26451fe3b9a2902b9ae5aef2a9

alexjo2144 commented:
Thanks David, looks like that worked. Do you want me to cherry-pick that over or would you rather merge it separately?

Also, there may be a few schemas in Glue that didn't get cleaned up properly, see the typo fix in the last fixup. They'd all start with `iceberg_smoke_test_`.  The other tests all look like they should be cleaning themselves up properly now.

electrum commented:
@alexjo2144 please cherry pick to this PR so that we can get this merged

alexjo2144 commented:
Cherry-picked and squashed, thanks

findepi commented:
@alexjo2144 please review the build results.

findepi commented:
(squashde)

mosabua commented:
Hm .. very interesting @findepi .. thanks for the link. Its a bit unfortunate that catalog type is overloaded as a term in Iceberg .. but I guess that is out of our control...

alexjo2144 commented:
> Its a bit unfortunate that catalog type is overloaded as a term in Iceberg

Agreed

@findepi @electrum builds are green! Can we go ahead and merge this in?
